docs/proposals/multi-tenant-networking.md

```diff
@@ -0,0 +1,194 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/multi-tenant-networking.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Multi-tenant networking
+
+## Abstract
+
+A proposal of multi-tenant networking by network providers.
+
+## Motivation
+
+The current kubernetes network design works great for connecting containers and building services, but it doesn’t work on many use-cases, such as container network isolation, ip address overlapping and so on. So this proposal adds network provider which adds multi-tenant networking for kubernetes.
+
+Goals of this design:
+
+* Add multi-tenant networking to support network isolation and ip address overlapping between containers
+* Support integration with SDN applications and external IPAMs by network provider
+* Works togather with current network model (if there is no network associated with the namespace, then pod networks of this namespace will work on current network model)
+
+## Use cases
+
+Multi-tenant networking is extremely useful for following use cases:
+
+* Network isotion between container networks
+* Manage kubernetes pod IPs by SDN applications
+* Integration with current IPAM tools
+
+## Proposed design
+
+To support multi-tenant networking, there are some changes to current kubernetes network design:
+
+* Instead of full communication between all containers, containers only on the same network are fullly communicated in this network model. Containers on different networks are not communicated directly, but they can communicate with each other via external endpoints.
+* Pods belongs to different network may overlap ip addresses. This is ok since networks are isolated.
+* Service ips are not virtual with network provider. They are allocated by network provider and load-balanced to specified pods.
+* Kube-proxy is network-aware, it only load-balances services whose namespace is with no network associated.
```

Comment:
The load balancing needs *some* integration with the network provider, to ensure that it can't be used to bypass isolation. (Eg, the problem we initially had in OpenShift; service isolation could not be implemented on the destination end, because the connection would always be masqueraded at that point so we couldn't check the actual source IP. But implementing the filtering on the source end meant that you could spoof things by creating a service in your own namespace but then manually adding an endpoint IP in another namespace. Current workaround: let the network plugin filter the list of endpoints before kube-proxy sees it.)